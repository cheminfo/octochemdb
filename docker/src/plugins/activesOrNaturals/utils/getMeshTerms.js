import debugLibrary from '../../../utils/Debug.js';

/**
 * @description Get the mesh terms and dbRef for a given cid inside the pubmeds collection
 * @param {Array} cids CIDs
 * @param {*} collection : PUBMEDS collection
 * @param {*} connection MongoDB connection
 * @returns {Promise} returns an object {meshTerms: array, dbRefs: array}
 */

export async function getMeshTerms(cids, collection, connection) {
  const debug = debugLibrary('getMeshTerms');
  // get id from dbRef
  let compoundIds = [];
  for (let cid of cids) {
    compoundIds.push(Number(cid.$id));
  }
  try {
    const result = await collection
      .aggregate([
        {
          $project: {
            _id: 1,
            data: 1,
            compounds: {
              $map: {
                input: '$data.compounds',
                as: 'compound',
                in: '$$compound.$id',
              },
            },
          },
        },
        { $match: { compounds: { $in: compoundIds } } },
        { $limit: 2 },
      ])
      .toArray();
    let uniqueMeshTerms = {};
    let pmids = [];
    for (let doc of result) {
      if (doc.data?.meshHeadings) {
        for (let meshHeading of doc.data.meshHeadings) {
          if (meshHeading.descriptorName) {
            const descriptorName = meshHeading.descriptorName.toLowerCase();
            uniqueMeshTerms[descriptorName] = true;
          }
        }
        pmids.push(doc._id);
      }
    }

    const counterPmids = pmids.length;
    return {
      meshTermsForCid: Object.keys(uniqueMeshTerms),
      pmIds: pmids,
      counterPmids,
    };
  } catch (error) {
    if (connection) {
      await debug.fatal(error, {
        collection: collection.collectionName,
        connection,
      });
    }
  }
}
