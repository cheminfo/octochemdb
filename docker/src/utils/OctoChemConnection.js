import delay from 'delay';
import dotenv from 'dotenv';
import md5 from 'md5';
import { MongoClient } from 'mongodb';

import debugLibrary from './Debug.js';

dotenv.config();
const debug = debugLibrary('OctoChemConnection');

export function OctoChemConnection() {
  this.client = new MongoClient(process.env.MONGODB_URL, {
    keepAlive: true,
    connectTimeoutMS: 6 * 60 * 60 * 1000,
    socketTimeoutMS: 6 * 60 * 60 * 1000,
  });
}

OctoChemConnection.prototype.close = async function close() {
  if (this.connection) return this.connection.close();
  return undefined;
};

OctoChemConnection.prototype.getCollectionNames =
  async function getCollectionNames() {
    return (await (await this.getDatabase()).listCollections().toArray()).map(
      (entry) => entry.name,
    );
  };

OctoChemConnection.prototype.getCollection = async function getCollection(
  collectionName,
) {
  return (await this.getDatabase()).collection(collectionName);
};

OctoChemConnection.prototype.getImportationLog =
  async function getImportationLog(options) {
    const { collectionName, sources, startSequenceID } = options;
    const logsCollection = await this.getImportationLogsCollection();
    const sourcesHash = md5(JSON.stringify(sources));
    const _id = sourcesHash;
    let logs = await logsCollection.find({ _id }).next();
    if (logs === null) {
      logs = {
        _id,
        collectionName,
        sources,
        sourcesHash,
        dateStart: Date.now(),
        dateEnd: Date.now(),
        startSequenceID,
        endSequenceID: 0,
        status: 'updating',
      };
      await logsCollection.insertOne(logs);
    }
    return logs;
  };
OctoChemConnection.prototype.updateImportationLog =
  async function updateImportationLog(logs) {
    const collection = await this.getImportationLogsCollection();
    logs.dateEnd = Date.now();
    await collection.replaceOne({ _id: logs._id }, logs);
  };

OctoChemConnection.prototype.getImportationLogsCollection =
  async function getCollection() {
    return (await this.getDatabase()).collection('importationLogs');
  };

OctoChemConnection.prototype.getAdminCollection =
  async function getAdminCollection() {
    return this.getCollection('admin');
  };

OctoChemConnection.prototype.setProgress = async function setProgress(
  progress,
) {
  const collection = await this.getAdminCollection();
  await collection.replaceOne({ _id: progress._id }, progress);
};

OctoChemConnection.prototype.getProgress = async function getProgress(
  collectionName,
) {
  const adminCollection = await this.getAdminCollection();
  const _id = `${collectionName}_progress`;
  let progress = await adminCollection.find({ _id }).next();
  if (progress === null) {
    progress = {
      _id,
      state: 'updating',
      seq: 0,
      dateStart: Date.now(),
      dateEnd: 0,
    };
    await adminCollection.insertOne(progress);
  }
  return progress;
};

OctoChemConnection.prototype.getDatabase = async function getDatabase() {
  while (true) {
    try {
      await this.init();
      break;
    } catch (e) {
      debug('Connection to mongo failed, waiting 5s');
      debug(e.stack);
      await delay(5000);
    }
  }
  return this.connection.db(process.env.MONGO_DB_NAME);
};

OctoChemConnection.prototype.init = async function init() {
  if (this.connection) return;

  debug(`Trying to connect to: ${process.env.MONGODB_URL}`);

  this.connection = await this.client.connect();
  debug('Got DB connection');
};
