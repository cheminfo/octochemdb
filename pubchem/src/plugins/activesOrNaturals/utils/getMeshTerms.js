import Debug from '../../../utils/Debug.js';

import { getMeshTermsType } from './getMeshTermsType.js';
/**
 * @description Get the mesh terms and dbRef for a given cid inside the pubmeds collection
 * @param {*} cid CID
 * @param {*} connection MongoDB connection
 * @returns {Promise} returns an array of objects {meshTerm: array,meshTermsType:array, dbRef: object}
 */
export async function getMeshTerms(cid, collection, connection) {
  const debug = Debug('getMeshTerms');
  try {
    const cursor = await collection.find({
      'data.cids': { $in: [cid] },
    });
    let results = [];
    while (await cursor.hasNext()) {
      const doc = await cursor.next();
      if (doc.data.meshHeadings) {
        let dbRef = { $ref: 'pubmeds', $id: doc._id };
        let meshTerms = [];
        let meshTermsType = [];
        for (let meshHeading of doc.data.meshHeadings) {
          let type = await getMeshTermsType(meshHeading.descriptorName);
          if (type !== 'other') {
            meshTermsType.push(type);
          }
          meshTerms.push(meshHeading.descriptorName);
        }

        let result = {};
        if (meshTerms.length > 0) {
          result.meshTerms = meshTerms;
        }
        if (meshTermsType.length > 0) {
          result.meshTermsType = meshTermsType;
        }
        result.dbRef = dbRef;
        results.push(result);
      }
    }
    return results;
  } catch (error) {
    if (connection) {
      debug(error, { collection: collection.collectionName, connection });
    }
  }
}
