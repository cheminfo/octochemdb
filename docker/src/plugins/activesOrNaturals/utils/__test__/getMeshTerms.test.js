import delay from 'delay';
import { test, expect } from 'vitest';

import { OctoChemConnection } from '../../../../utils/OctoChemConnection.js';
import { getMeshTerms } from '../getMeshTerms.js';

test(
  'Aggregation ActivesOrNaturals',
  async () => {
    const connection = new OctoChemConnection();
    let colllectionList = await connection.getCollectionNames();
    while (!colllectionList.includes('pubmeds')) {
      await delay(1000);
      colllectionList = await connection.getCollectionNames();
    }
    const collection = await connection.getCollection('pubmeds');
    let cids = [
      { $ref: 'compounds', $id: 19342306 },
      { $ref: 'compounds', $id: 19342308 },
    ];

    let result = await getMeshTerms(cids, collection, connection);

    expect(result).toMatchSnapshot();
    await connection.close();
  },
  { timeout: 100000 },
);
