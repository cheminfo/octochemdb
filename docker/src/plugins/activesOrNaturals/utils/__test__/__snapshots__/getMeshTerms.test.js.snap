// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Aggregation ActivesOrNaturals 1`] = `
{
  "counterPmids": 5,
  "meshTermsForCid": [
    "abdominal injuries",
    "alcoholic intoxication",
    "cardiac surgical procedures",
    "cardiac tamponade",
    "heart atria",
    "heart block",
    "heart injuries",
    "humans",
    "multiple trauma",
    "nephrectomy",
    "resuscitation",
    "thoracic surgery",
    "isotopes",
    "medicine",
    "radioisotopes",
    "research",
    "antipsychotic agents",
    "benzodiazepines",
    "bipolar disorder",
    "drug therapy, combination",
    "olanzapine",
    "pirenzepine",
    "randomized controlled trials as topic",
    "treatment failure",
    "animals",
    "bile acids and salts",
    "cells, cultured",
    "deoxycholic acid",
    "gtp-binding protein alpha subunits, gi-go",
    "glycogen synthase",
    "hepatocytes",
    "male",
    "phosphatidylinositol 3-kinases",
    "proto-oncogene proteins c-akt",
    "rats",
    "rats, sprague-dawley",
    "signal transduction",
    "taurodeoxycholic acid",
    "activities of daily living",
    "adult",
    "body composition",
    "disability evaluation",
    "female",
    "gastric bypass",
    "laparoscopy",
    "longitudinal studies",
    "middle aged",
    "motor activity",
    "muscle strength",
    "obesity, morbid",
    "surveys and questionnaires",
    "task performance and analysis",
    "weight loss",
  ],
  "pmIds": [
    14248046,
    14248047,
    12918000,
    17200418,
    19342308,
  ],
}
`;
